Source/Quartz.Plugins.RecentHistory/Impl/SqlServer/SqlServerExecutionHistoryStore.cs

```diff
@@ -0,0 +1,308 @@
+                    sqlCommand.Parameters.AddWithValue("@SchedulerInstanceId", entry.SchedulerInstanceId);
+                    sqlCommand.Parameters.AddWithValue("@SchedulerName", entry.SchedulerName);
+                    sqlCommand.Parameters.AddWithValue("@Job", entry.Job);
+                    sqlCommand.Parameters.AddWithValue("@Trigger", entry.Trigger);
+                    sqlCommand.Parameters.AddWithValue("@ScheduledFireTimeUtc", entry.ScheduledFireTimeUtc != null ? (object)entry.ScheduledFireTimeUtc : DBNull.Value);
+                    sqlCommand.Parameters.AddWithValue("@ActualFireTimeUtc", entry.ActualFireTimeUtc);
+                    sqlCommand.Parameters.AddWithValue("@Recovering", entry.Recovering);
+                    sqlCommand.Parameters.AddWithValue("@Vetoed", entry.Vetoed);
+                    sqlCommand.Parameters.AddWithValue("@FinishedTimeUtc", entry.FinishedTimeUtc != null ? (object)entry.FinishedTimeUtc : DBNull.Value);
+                    sqlCommand.Parameters.AddWithValue("@ExceptionMessage", entry.ExceptionMessage != null ? (object)entry.ExceptionMessage : DBNull.Value);
+
+                    await sqlCommand.ExecuteNonQueryAsync();
+                }
+            }
+        }
+
+        public async Task Purge()
+        {
+            using (var sqlConnection = new SqlConnection(ConnectionString))
+            {
+                await sqlConnection.OpenAsync();
+
+                string commandText = $"DELETE FROM {GetTableName(C.TableExecutionHistoryEntries)} WHERE {C.ColumnActualFireTimeUtc} < @PurgeThreshold";
+                using (var sqlCommand = new SqlCommand(commandText, sqlConnection))
+                {
+                    sqlCommand.Parameters.AddWithValue("@PurgeThreshold", DateTime.UtcNow.Subtract(TimeSpan.FromMinutes(EntryTTLInMinutes)));
+
+                    await sqlCommand.ExecuteNonQueryAsync();
+                }
+            }
+        }
+
+        public async Task<IEnumerable<ExecutionHistoryEntry>> FilterLastOfEveryJob(int limitPerJob)
+        {
+            return await FilterLastOf(C.ColumnJob, limitPerJob);
+        }
+
+        public async Task<IEnumerable<ExecutionHistoryEntry>> FilterLastOfEveryTrigger(int limitPerTrigger)
+        {
+            return await FilterLastOf(C.ColumnTrigger, limitPerTrigger);
+        }
+
+        protected async Task<IEnumerable<ExecutionHistoryEntry>> FilterLastOf(string columnName, int limit)
+        {
+            string query =
+                $"WITH SELECTION AS ( \n" +
+                $"	SELECT *, \n" +
+                $"	ROW_NUMBER() OVER(PARTITION BY {columnName} ORDER BY {C.ColumnActualFireTimeUtc} DESC) AS ROW_KEY \n" +
+                $"	FROM {GetTableName(C.TableExecutionHistoryEntries)} \n" +
+                $"	WHERE {C.ColumnSchedulerName} = @SchedulerName \n" +
+                $") \n" +
+                $"SELECT * \n" +
+                $"FROM SELECTION \n" +
+                $"WHERE ROW_KEY <= {limit}";
+
```

Comment:
I noticed the original in memory implementation had
`.Select(x => x.OrderByDescending(y => y.ActualFireTimeUtc).Take(limitPerJob).Reverse())` within the group by.

I'm not sure if it's necessary or not, but if you wanted to keep it, you could add:

`$"ORDER BY {columnName}, ROW_KEY DESC";`

and that should perform the same Reverse.